Basic/DyanmicPrefabs/Assets/Scripts/01_Server authoritative dynamic prefab loading across peers/DynamicPrefabManager.cs

```diff
@@ -0,0 +1,476 @@
+        {
+            Debug.Log(nameof(StartHost));
+            m_ConnectAddress = connectAddress;
+            m_Port = port;
+            m_NetworkManager.NetworkConfig.ForceSamePrefabs = false;
+            var transport = m_NetworkManager.GetComponent<UnityTransport>();
+            transport.SetConnectionData(connectAddress, port);
+            m_NetworkManager.ConnectionApprovalCallback = ConnectionApprovalCallback;
+            m_NetworkManager.StartHost();
+        }
+
+        public override void OnDestroy()
+        {
+            UnloadAndReleaseAllDynamicPrefabs();
+            base.OnDestroy();
+        }
+
+        async void ReceiveServerToClientSetDisconnectReason_CustomMessage(ulong clientID, FastBufferReader reader)
+        {
+            reader.ReadValueSafe(out ConnectStatus status);
+
+            switch (status)
+            {
+                case ConnectStatus.Undefined:
+                    Debug.Log("Undefined");
+                    m_NetworkManager.Shutdown();
+                    break;
+                case ConnectStatus.ClientNeedsToPreload:
+                {
+                    m_NetworkManager.Shutdown();
+                    Debug.Log("Client needs to preload");
+                    reader.ReadValueSafe(out AddressableGUIDCollection addressableGUIDCollection);
+                    await LoadDynamicPrefabs(addressableGUIDCollection);
+                    Debug.Log("Restarting client");
+                    StartClient(m_ConnectAddress, m_Port);
+                }
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+
+            OnConnectionStatusReceived?.Invoke(status, reader);
+        }
+        
+        /// <summary>
+        /// Sends a DisconnectReason to the indicated client. This should only be done on the server, prior to disconnecting the client.
+        /// </summary>
+        /// <param name="clientID"> id of the client to send to </param>
+        /// <param name="status"> The reason for the upcoming disconnect.</param>
+        /// <param name="addressableGUIDCollection"></param>
+        void SendServerToClientSetDisconnectReason(ulong clientID, ConnectStatus status, AddressableGUIDCollection addressableGUIDCollection)
+        {
+            int guidCollectionSize = addressableGUIDCollection.GetSizeInBytes();
+            
+            var writer = new FastBufferWriter(sizeof(ConnectStatus) + guidCollectionSize, Allocator.Temp);
+            writer.WriteValueSafe(status);
+            writer.WriteValueSafe(addressableGUIDCollection);
+            
+            m_NetworkManager.CustomMessagingManager.SendNamedMessage(nameof(ReceiveServerToClientSetDisconnectReason_CustomMessage), clientID, writer);
+        }
+        
+        void ConnectionApprovalCallback(NetworkManager.ConnectionApprovalRequest request, NetworkManager.ConnectionApprovalResponse response)
```

Comment:
@fernando-cortez will need to split this up in a state machine like in boss room.